project13/ViewController.swift

```diff
@@ -4,16 +4,191 @@
+        blurEffectView.alpha = 0.9
+        blurEffectView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
+        return blurEffectView
+    }()
+    
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        
+        let newTitle: String = "InstaFilter"
+        title = newTitle
+        
+        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(importPicture))
+
+        currentFilter = CIFilter(name: "CITwirlDistortion")
+        
+    }
+    @objc private func importPicture() {
+        startLoadingAnimation()
+        DispatchQueue.main.async { [weak self] in
+            guard let self = self else { return }
+            let picker = UIImagePickerController()
+            picker.allowsEditing = true
+            picker.delegate = self
+            self.present(picker, animated: true) {
+                self.stopLoadingAnimation()
+            }
+        }
+    }
+    
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        
+        guard let image = info[.editedImage] as? UIImage else { return }
+        dismiss(animated: true)
+        currentImage = image
+        guard let currentImage = currentImage else { return }
+        
+        let beginImage = CIImage(image: currentImage)
+        currentFilter.setValue(beginImage, forKey: kCIInputImageKey)
+        applyProcessing()
+        stopLoadingAnimation()
+        
+    }
+    
+    @IBAction func changeFilter(_ sender: UIButton) {
+        let alertController = UIAlertController(title: "Choose filter", message: nil, preferredStyle: .actionSheet)
+        CIFilterName.allCases.forEach { filter in
+                  let action = UIAlertAction(title: filter.stringName, style: .default, handler: { _ in
+                      self.set(filter: filter)
+                  })
+                  alertController.addAction(action)
+              }
+        if let popoverController = alertController.popoverPresentationController {
+            popoverController.sourceView = sender
+            popoverController.sourceRect = sender.bounds
+        }
+        present(alertController, animated: true)
+    }
+    
+    @IBAction func save(_ sender: UIButton) {
+        guard let image = imageView.image else {
+            let alertController = UIAlertController(title: "Error", message: "Please selct an image", preferredStyle: .actionSheet)
+            alertController.addAction(UIAlertAction(title: "Ok", style: .destructive))
+            present(alertController, animated: true)
+            return
+        }
+        UIImageWriteToSavedPhotosAlbum(image, self, #selector(image(_:didFinishSavingWithError: contextInfo:)),nil)
+    }
+    
+    @IBAction func intensityChanger(_ sender: Any) {
+        applyProcessing()
+    }
```

Comment:
I found the name for this method weird because it performs an action but it doesn't start with a verb. A better name would be `changeFilterIntensity(_:)`. Here is an Apple doc about naming in iOS development: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html. This article also seems interesting: https://betterprogramming.pub/naming-conventions-in-swift-4b7ca5eed4d2#893b.

Another issue is that this method doesn't do actually anything. You could make your `applyProcessing()` an `@IBAction` and then the same would happen. Regardless, I believe you should keep it, but I also think that `applyProcessing` should receive the value of the intensity in order to be more isolated (i.e. separation of concerns). This would read like so:

```suggestion
    @IBAction func changeFilterIntensity(_ sender: UISlider) {
        applyProcessing(filterIntensity: sender.value)
    }
```

Please notice I replaced the sender type to avoid casting, but you could also do this:

```suggestion
    @IBAction func changeFilterIntensity(_ sender: Any) {
        guard let slider = sender as? UISlider else { return }
        applyProcessing(filterIntensity: sender.value)
    }
```

Reply:
Excellent point. I will keep those links on my favorites bar!